wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,118 @@
+<?php
+/**
+ * Add/update/retrieve User progress
+ * 
+ */
+class TapestryUserController {
+
+    /** 
+     * Update User's video progress for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+     * @param type @progressData json data for progress of each node
+    */
+    public function updateProgress($userId, $postId, $progressData) {
+        $this->checkUserAndPostId($userId, $postId);  
+
+        if ($this->isJson($progressData)) {
+            $progressData = json_decode($progressData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->updateUserProgress($userId, $postId, $progressData);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+    */
+    public function getProgress($userId, $postId) {
+        $this->postExists($postId);
+        $this->checkUserAndPostId($userId, $postId);  
+        return $this->getUserProgress($userId, $postId);
+    }
+
+    /** 
+     * Update User's h5p video setting for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+    */
+    public function updateH5PSetting($userId, $postId, $h5pSettingsData) {
+        $this->checkUserAndPostId($userId, $postId);  
+
+        if ($this->isJson($h5pSettingsData)) {
+            $h5pSettingsData = json_decode($h5pSettingsData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->updateUserH5PSetting($userId, $postId, $h5pSettingsData);
+    }
+
+    /** 
+     * Get User's h5p video setting for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+    */
+    public function getH5PSetting($userId, $postId) {
+        $this->postExists($postId);
+        $this->checkUserAndPostId($userId, $postId);  
+        return $this->getUserH5PSetting($userId, $postId);
+    }
+
+    private function updateUserProgress($userId, $postId, $progressData) {
+        update_user_meta($userId, 'progress_data_' . $postId, $progressData);
```

Comment:
Please prefix this meta key and all other meta keys with `tapestry_` to avoid conflicts. You may want to use a format like `tapestry_{tapestryId}_progress_node_{nodeMetaId}` or like `tapestry_progress_node_{nodeMetaId}` to store the progress for a given node

Reply:
Saved key as tapestry_{tapestryId}_progress_node_{nodeMetaId}